### ISSUE TITLE
run-ios bug: Try/catch removed from `warnAboutManuallyLinkedLibs.ts` causing exceptions to propagate/crash

### ISSUE THREAD

**Issue by Saadnajmi** [react-native-community/cli#1435]
## Environment

```
System:
  OS: macOS 11.4
  CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  Memory: 921.95 MB / 16.00 GB
  Shell: 3.2.57 - /bin/bash
Binaries:
  Node: 16.4.0 - /usr/local/bin/node
  Yarn: 1.22.10 - /usr/local/bin/yarn
  npm: 7.18.1 - /usr/local/bin/npm
  Watchman: 4.9.0 - /usr/local/bin/watchman
Managers:
  CocoaPods: 1.10.1 - /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
  Android SDK:
    API Levels: 28, 29, 30
    Build Tools: 28.0.3, 29.0.3, 30.0.1
    System Images: android-30 | Google APIs Intel x86 Atom
    Android NDK: Not Found
IDEs:
  Android Studio: 4.1.2 4.1.2
  Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
Languages:
  Java: 11.0.1 - /usr/bin/javac
npmPackages:
  @react-native-community/cli: Not Found
  react: 17.0.1 => 17.0.1 
  react-native: 0.64.1 => 0.64.1 
  react-native-macos: ^0.63.0 => 0.63.36 
npmGlobalPackages:
  *react-native*: Not Found
```

## Description

This refers to https://github.com/microsoft/react-native-test-app/issues/375

I was working on moving a community module to use [React Native Test App](https://github.com/microsoft/react-native-test-app/) to manage the test app of a community module (https://github.com/react-native-menu/menu), and ran into an error running `run-ios`. After investigation, it seems the issue happened upstream beginning in https://github.com/react-native-community/cli/pull/1270 .

A try-catch was removed from `packages/platform-ios/src/link/warnAboutManuallyLinkedLibs.ts`, causing the exception to bubble up and crash. This seems like an error, because an Android fix shouldn't have broken something downstream for iOS.

## Reproducible Demo

A little bit contrived since I reproed this using a downstream module, but:
```
git clone git@github.com:react-native-menu/menu.git
cd menu
mv example/ example2/ # This is to just move the existing test app out of the way temporarily
yarn add react-native-test-app --dev
yarn init-test-app
# The options I chose
# ✔ What is the name of your test app? … MenuExample
# ✔ Which platforms do you need test apps for? › Android, iOS
# ✔ Where should we create the new project? … example
yarn
yarn build:ios
pod install --project-directory=ios
yarn ios
```

**Comment by tido64** [comment]
I have a draft PR with a proposed solution ☝️

If you can, please give it a try and let me know how it goes.

cc @grabbou, @thymikee

**Comment by grabbou** [comment]
For this particular thing to work, we need to do two things:

1/ Update the config to properly detect a Podfile [feature]
2/ Teach `run-ios` about a non-standard source dir [bug]

It is a mix of features and bugs, I will try to be as explicit as possible, the list of changes is really minimal.

<hr />

### Ad. 1

First of all, we need to set a proper path to a project. Since `.xcodeproj` doesn't live within the project directory, but inside node_modules, we need to update the configuration to reflect that. 

Luckily, this can be done right now with a configuration:

```js
module.exports = {
  project: {
    ios: {
      project: "node_modules/react-native-test-app/ios/ReactTestApp.xcodeproj",
    },
  },
};
```

When we do it, we can see the output of `react-native config` looks as follows:

```json
  "ios": {
      "sourceDir": "/Users/grabbou/Repositories/menu/example/node_modules/react-native-test-app/ios",
      "folder": "/Users/grabbou/Repositories/menu/example",
      "pbxprojPath": "/Users/grabbou/Repositories/menu/example/node_modules/react-native-test-app/ios/ReactTestApp.xcodeproj/project.pbxproj",
      "podfile": null,
      "podspecPath": null,
      "projectPath": "/Users/grabbou/Repositories/menu/example/node_modules/react-native-test-app/ios/ReactTestApp.xcodeproj",
      "projectName": "ReactTestApp.xcodeproj",
      "libraryFolder": "Libraries",
      "sharedLibraries": [],
      "plist": [],
      "scriptPhases": []
    },
```

As you can see, `sourceDir` points to a folder within `node_modules`, where the project is located. It is not a surprise, since it is just a `path.dirname` on a supplied `project` value.

However, `podfile`, and `podspecPath` are `null`, this is because we look for them within `sourceDir`. **I believe this is a right default heuristic**.

Rather than fighting with direction mechanisms and globs, I would just **let people set a path to a podfile**. 

**In other words, make it possible to pass the "podfile" as a part of config**:

```js
module.exports = {
  project: {
    ios: {
      project: "node_modules/react-native-test-app/ios/ReactTestApp.xcodeproj",
      podfile: "ios/Podfile"
    },
  },
};
```

which will be useful in the future anyway. 

**This is needed to make autolinking to work**. However, `run-ios` is still not going to work. For a fix and explanation why, please see the next chapter.

<hr />

### Ad. 2

When you do the following configuration:

```js
module.exports = {
  project: {
    ios: {
      project: "node_modules/react-native-test-app/ios/ReactTestApp.xcodeproj",
    },
  },
};
```

and do `run-ios`, you're going end up with the following:

```
error Could not find Xcode project files in "ios" folder.
info Run CLI with --verbose flag for more details.
```

This is because, `react-native run-ios` doesn't use config. And it's a bug. By default, `react-native run-ios` assumes that `projectPath` is `ios`. You can change that value by passing a `--projectPath` flag.

First of all, it should be named `sourceDir`, since it's expected to be a directory, not a path to a file. Second, it doesn't respect the value from config - in other words - if your `xcodeproj` lives in a different location (like we just set in custom config), it is going to fail.

To fix it, it is very simple. 

Replace the `default` in the following:

```js
{
      name: '--project-path <string>',
      description:
        'Path relative to project root where the Xcode project ' +
        '(.xcodeproj) lives.',
      default: 'ios',
    },
```

to 

```js
default: (ctx: Config) => ctx.project.ios ? ctx.project.ios.sourceDir : 'ios',
```

and viola, the `projectPath` is now automatically derived from the configuration. When users run `react-native --help`, they will see the default value (real one).

<hr />

TL;DR, we need to do two changes:
- accept `podfile` as a custom configuration
- fix `run-ios` not reading `sourceDir` from configuration by default

and I believe everything should work (at least `run-ios`, tested as I was writing this).

**Comment by tido64** [comment]
Thanks for the detailed explanation, @grabbou. There is one thing I don't quite understand though. It sounds like the only piece of information that we need from the user is where the `Podfile` is. `.xcworkspace` is _always_ generated next to it, so we can derive both `sourceDir` and `project`. What do we need everything else for? It doesn't look like we need the `.xcodeproj` for anything. Even if we do need it, we should be getting the path from `.xcworkspace`. Assuming that `.xcodeproj` is always next to Podfile is just wrong. It could be anywhere as it can be shared between multiple projects.

**Comment by grabbou** [comment]
> What do we need everything else for?

We don't need it, it's been used for legacy link. That's why I don't want to derive anything from `podfile` location, since in the next version that we should get started on:

> It sounds like the only piece of information that we need from the user is where the Podfile is

As you have noticed yourself! It will become the only thing (well, except for maybe scriptPhases) that is needed.

<hr />

I agree that we should not assume that these exist next to each other, but again, the default values were meant to satisfy the most popular setup (aka the default one). This needs a general clean up, so I feel like adding an override is kinda safe when defaults are broken, while we work on the new thing.

**Comment by grabbou** [comment]
If you look at the `next` branch

https://github.com/react-native-community/cli/blob/next/packages/platform-ios/src/config/index.ts#L41-L45

only three properties are used. I didn't remove the `project` check back then, but we can safely replace that with a `podfile` (like in your PR).

**Comment by grabbou** [comment]
For context, we used to rely on an `.xcodeproj`. Because projects were all different, we started with a detection mechanism (`glob`) to eliminate naming collisions when people would use something different than `iOS`, at a different level too.

In retrospection, I don't think we necessarily need to support non-standard locations. If someone's using something different than `iOS` or is having a non-standard project structure, overrides should be the way to go, instead of a complex heuristics, like we sometimes have.

When CocoaPods support arrived, it was first optional, then became the default. Since it was "by default" next to `xcodeproj`, we used the `sourceDir`, which is derived from an `xcodeproj`. Almost from the very beginning, the presence of a project was dictating that it was a valid React Native project. As it can be observed in this thread, it is no longer the case and that is yet another reason to upgrade this code.

**Comment by tido64** [comment]
Thanks for the context. It makes total sense given the history. Is manual linking still used with version 5.x and 6.x?

**Comment by grabbou** [comment]
Update this is being addressed in https://github.com/react-native-community/cli/pull/1537/commits/596d849d30c07ec056c648afe8d17460afd66d8b#diff-501be5ad14353a5d85efd5614d2571d96b92ba33ad8c9fadc40aed0b58de87ebR115 where I have made Podfile detection the source of truth.
